iroh-net/src/magicsock/peer_map/endpoint.rs

```diff
@@ -199,68 +199,86 @@ impl Endpoint {
-            // we have a valid address: use it!
             best_addr::State::Valid(best_addr) => {
-                trace!(addr = %best_addr.addr, latency = ?best_addr.latency, "best_addr is set and valid, use best_addr only");
-                (Some(best_addr.addr), None, false)
+                // If we have a valid address we use it.
+                trace!(addr = %best_addr.addr, latency = ?best_addr.latency,
+                       "best_addr is set and valid, use best_addr only");
+                (Some(best_addr.addr), None)
             }
-            // we have an outdated address: use it, but use derp as well.
             best_addr::State::Outdated(best_addr) => {
-                trace!(addr = %best_addr.addr, latency = ?best_addr.latency, "best_addr is set but outdated, use best_addr and derp");
-                (Some(best_addr.addr), self.derp_url(), true)
+                // If the address is outdated we use it, but send via derp at the same time.
+                // We also send disco pings so that it will become valid again if it still
+                // works (i.e. we don't need to holepunch again).
+                trace!(addr = %best_addr.addr, latency = ?best_addr.latency,
+                       "best_addr is set but outdated, use best_addr and derp");
+                (Some(best_addr.addr), self.derp_url())
             }
-            // we have no best address: use a random candidate if available, and derp as backup.
             best_addr::State::Empty => {
+                // No direct connection has been used before.  If we know of any possible
+                // candidate addresses, randomly try to use one while also sending via derp
+                // at the same time.  If we use a candidate address also send disco ping
+                // message to it.
                 let addr = self
                     .direct_addr_state
                     .keys()
                     .choose_stable(&mut rand::thread_rng())
                     .map(|ipp| SocketAddr::from(*ipp));
                 trace!(udp_addr = ?addr, "best_addr is unset, use candidate addr and derp");
-                let should_ping = addr.is_some()
-                    || self
-                        .derp_url
-                        .as_ref()
-                        .map(|(_r, state)| state.needs_ping(now))
-                        .unwrap_or(false);
-                (addr, self.derp_url(), should_ping)
+                (addr, self.derp_url())
             }
         }
     }
 
     /// Update our best_addr (if empty) with the candidate udp addr with the lowest latency.
     fn assign_best_addr_from_candidates_if_empty(&mut self) {
+        // TODO(flub): I'm unsure how this function can be responsible to ever fill in the
+        // best_path.  We'd have to have a path that has a known latency, yet no pong should
+        // have been received on that path as otherwise the pong handler would have already
+        // selected the path.
```

Comment:
the point of this function is to select the best guess in the face of missing (potentially no latency) information

Reply:
yes, but this should not happen.  When we receive a pong it should be promoted to best_addr.

However I have seen this happening: when pasts pings timeout they clear the best_addr state.  Even though there was a more recent successful ping.  That bug needs to be fixed, but this papered over it.

I don't think I'll remove this though, that's a bit risky probably as this is a good fallback.